### PR TITLE
Add Config.keep_untouched for custom descriptors support

### PR DESCRIPTION
**Pull request by MrMrRobat**
## Change Summary
This will allow to extend default `TYPE_BLACKLIST` during class creation. 

It's very helpful if we want use custom descriptors for our class (like [cached_property](https://github.com/pydanny/cached-property), etc) and don't want pydantic to break them.

```py
from pydantic import BaseModel

class _ClassPropertyDescriptor:
    __slots__ = ('getter', )

    def __init__(self, getter):
        self.getter = getter

    def __get__(self, instance, owner):
        return self.getter(owner)

classproperty = _ClassPropertyDescriptor

class Model(BaseModel):
    class Config:
        keep_untouched = (classproperty, )

    @classproperty
    def class_name(cls) -> str:
        return cls.__name__

print(Model.class_name)  # Model
print(Model().class_name)  # Model
```

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `HISTORY.rst` has been updated
  * if this is the first change since a release, please add a new section
  * include the issue number or this pull request number `#<number>`
  * include your github username `@<whomever>`

### CODE REVIEW

File: pydantic/main.py [hunk 1]
```diff
@@ -214,10 +215,11 @@ def __new__(mcs, name, bases, namespace):
                         config=config,
                     )
 
+            type_blacklist = TYPE_BLACKLIST + config.keep_untouched
```
Comment (samuelcolvin) [inline pydantic/main.py]: please rename `TYPE_BLACKLIST` to `UNTOUCHED_TYPES` to be consistent with config.

File: pydantic/validators.py [hunk 2]
```diff
@@ -474,7 +474,7 @@ def find_validators(  # noqa: C901 (ignore complexity)
     if config.arbitrary_types_allowed:
         yield make_arbitrary_type_validator(type_)
     else:
-        raise RuntimeError(f'no validator found for {type_}')
+        raise RuntimeError(f'no validator found for {type_}\n see keep_untouched or arbitrary_types_allowed in Config')
```
Comment (samuelcolvin) [inline pydantic/validators.py]:
```suggestion
        raise RuntimeError(f'no validator found for {type_} see `keep_untouched` or `arbitrary_types_allowed` in Config')
```

File: HISTORY.rst [hunk 3]
```diff
@@ -8,6 +8,7 @@ v0.31 (unreleased)
 * better support for floating point `multiple_of` values, #652 by @justindujardin
 * fix schema generation for ``NewType`` and ``Literal``, #649 by @dmontagu
 * add documentation for Literal type, #651 by @dmontagu
+* add Config.keep_untouched for custom descriptors support, #679 by @MrMrRobat
```
Comment (samuelcolvin) [inline HISTORY.rst]:
```suggestion
* add ``Config.keep_untouched`` for custom descriptors support, #679 by @MrMrRobat
```

File: docs/index.rst [hunk 4]
```diff
@@ -762,6 +762,7 @@ Options:
     details.
 :orm_mode: allows usage of :ref:`ORM mode <orm_mode>`
 :alias_generator: callable that takes field name and returns alias for it
+:keep_untouched: custom types of model attributes (e. g. descriptors) that won't change during model creation
```
Comment (samuelcolvin) [inline docs/index.rst]:
```suggestion
:keep_untouched: set of types (e. g. descriptors) that won't change during model creation and won't be
  included in the model schemas.
```

Sorry, should have caught this the first time.

Comment (MrMrRobat) [reply docs/index.rst]: NP. But how about change 'set of types' to 'tuple of types'? I guess you meant 'set'  in general, but I still find this a bit confusing. What do you think?

Comment (samuelcolvin) [reply docs/index.rst]: yes of course, my mistake.

File: pydantic/main.py (hunk 1 above)
Comment (MrMrRobat) [reply pydantic/main.py]: Done :)

**Pull request merged**